fix: set wallet start scan height to birthday and not 0 (see issue #4807)

jorgeantonio21 opened the pull request:
Description
---
When wallet starts new UTXO scanning service it should start at the birthday. Currently, it starts at height 0. This PR addresses this issue

Motivation and Context
---
Addresses #4807.

How Has This Been Tested?
---

jorgeantonio21 committed:
first commit

SWvheerden commented on base_layer/wallet/src/utxo_scanner_service/utxo_scanner_task.rs:
```diff
@@ -325,7 +325,11 @@ where
     ) -> Result<BlockHeader, UtxoScannerError> {
         let tip_info = client.get_tip_info().await?;
-        let chain_height = tip_info.metadata.map(|m| m.height_of_longest_chain()).unwrap_or(0);
+        let wallet_birthday_height = self.get_birthday_header_height_hash(client).await?.height;
```
The old code is correct. 
If the wallet cant get the network tip from the base_node its connected to it needs to assume its 0. 
It also very unlikely that wallet wont have a tip, but has random height header?

jorgeantonio21 replied:
Great point ! You are right :)

SWvheerden reviewed (changes requested):
We need to implement this logic when the node selects its starting height for a scan, not the ending scan. 
The end scan should always be the network tip.

AaronFeickert commented on base_layer/wallet/src/utxo_scanner_service/utxo_scanner_task.rs:
```diff
@@ -66,6 +66,7 @@ use crate::{
     },
 };
 
+pub const BIRTHDAY_GENESIS_FROM_UNIX_EPOCH: u64 = 1640995200; // seconds since 2022-01-01 00:00:00 UTC
```
This shouldn't be a hard-coded value. It needs to match the [value](https://github.com/tari-project/tari/blob/43efc07a9713ebb498ca635603b97c34b561b1e2/base_layer/key_manager/src/cipher_seed.rs#L54) corresponding to the `CipherSeed` version in question (of which there is currently only one).

jorgeantonio21 replied:
I was actually playing around to test it. The best approach, imo, is to add a method to get the birthday date from unix epoch, like `get_birthday_time_from_unix()`.

jorgeantonio21 committed:
add changes

jorgeantonio21 committed:
add birthday from unix epoch util function

jorgeantonio21 committed:
refactor calculation of birthday epoch time

jorgeantonio21 committed:
correct tests

jorgeantonio21 committed:
Merge branch 'development' into ja-wallet-scans-at-birthday-not-height-0

SWvheerden commented on base_layer/wallet/src/utxo_scanner_service/utxo_scanner_task.rs:
```diff
@@ -699,9 +700,10 @@ where
-        let epoch_time = u64::from(birthday.saturating_sub(2)) * 60 * 60 * 24;
+        // Calculate the unix epoch time of two days before the wallet birthday.
+        // This is to avoid any weird time zone issues
+        let epoch_time = get_birthday_from_unix_epoch(birthday, 2u16);
```
I would go for 2 weeks, rather than 2 days. 
This is to go around re-org issues, we should rather error on the safe side/

SWvheerden reviewed (dismissed):
Looks good to me, I would just up the time to be longer to error on the safe side

jorgeantonio21 committed:
take 14 days back from birthday to account for reorgs

SWvheerden reviewed (dismissed):
utACK

AaronFeickert commented on base_layer/key_manager/src/lib.rs:
```diff
@@ -30,3 +31,7 @@ pub(crate) fn mac_domain_hasher<D: Digest + LengthExtensionAttackResistant>(
     DomainSeparatedHasher::<D, KeyManagerDomain>::new_with_label(label)
 }
+
+pub fn get_birthday_from_unix_epoch(birthday: u16, to_days: u16) -> u64 {
+    u64::from(birthday.saturating_sub(to_days)) * 24 * 60 * 60 + BIRTHDAY_GENESIS_FROM_UNIX_EPOCH
+}
```
Nit: recommend a function-level comment for this, to indicate its purpose. Might also be helpful to rename it in a way that indicates an extra offset is present and clearly describes the unit conversion.

jorgeantonio21 replied:
I added some documentation and add `_in_seconds` to the naming of the function

AaronFeickert commented on base_layer/wallet/src/utxo_scanner_service/utxo_scanner_task.rs:
```diff
@@ -697,9 +698,10 @@ where
-        // issues
-        let epoch_time = u64::from(birthday.saturating_sub(2)) * 60 * 60 * 24;
+        // Calculate the unix epoch time of two weeks (14 days) before the wallet birthday.
+        // This is to avoid any weird time zone issues
```
Time zone issues?

jorgeantonio21 replied:
Actually, the main reason is to avoid any possible reorgs. I updated the comment to convey this more precisely.

AaronFeickert commented on base_layer/wallet/tests/utxo_scanner.rs:
```diff
@@ -288,7 +288,7 @@ async fn test_utxo_scanner_recovery() {
 
     let cipher_seed = CipherSeed::new();
-    let birthday_epoch_time = u64::from(cipher_seed.birthday() - 2) * 60 * 60 * 24;
+    let birthday_epoch_time = get_birthday_from_unix_epoch(cipher_seed.birthday(), 14u16);
```
Nit: recommend a comment clearly indicating the units, since we switch between seconds and days.

jorgeantonio21 replied:
Definitely !

AaronFeickert reviewed (dismissed):
Just a few nits for clarity, particularly because of unit conversions between the `CipherSeed` birthday and Unix time.

Are there tests that exercise the expected behavior? Namely:
- a birthday corresponding to a block that is not the genesis block
- a birthday corresponding to the genesis block

If not, recommend adding them if feasible.

Otherwise, did not specifically test, but the changes to the math appear to be doing what you want.

AaronFeickert commented:
Any reason for the large changes to `Cargo.lock`? Seems like a lot.

AaronFeickert commented:
May be useful to include a couple of unit tests to hit the saturating subtraction cases.

jorgeantonio21 committed:
Merge branch 'development' into ja-wallet-scans-at-birthday-not-height-0

jorgeantonio21 committed:
add PR changes

jorgeantonio21 committed:
Merge branch 'development' into ja-wallet-scans-at-birthday-not-height-0

jorgeantonio21 committed:
add test birthday timestamp over chain

jorgeantonio21 commented:
Added unit tests and documentation that convey, more precisely, the points made by @AaronFeickert. I tried to make unit tests to be as expressive as possible. In particular, we use hardcoded values for second duration from specific dates. Also added a test on a mocked chain to check consistency of birthday and block timestamps, altogether.

jorgeantonio21 committed:
good practices
